File: iBurn/src/main/java/com/gaiagps/iburn/database/ArtDao.java
package com.gaiagps.iburn.database;

import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Update;

import java.util.List;

import io.reactivex.Flowable;

import static com.gaiagps.iburn.database.Art.TABLE_NAME;
import static com.gaiagps.iburn.database.PlayaItem.FAVORITE;
import static com.gaiagps.iburn.database.PlayaItem.LATITUDE;
import static com.gaiagps.iburn.database.PlayaItem.LONGITUDE;
import static com.gaiagps.iburn.database.PlayaItem.NAME;
import static com.gaiagps.iburn.database.PlayaItem.PLAYA_ID;

/**
 * Created by dbro on 6/8/17.
 */

@Dao
public interface ArtDao {

    @Query("SELECT * FROM " + TABLE_NAME + " ORDER BY " + NAME)
    Flowable<List<Art>> getAll();

    @Query("SELECT * FROM " + TABLE_NAME + " WHERE " + FAVORITE + " = 1")
    Flowable<List<Art>> getFavorites();

    @Query("SELECT * FROM " + TABLE_NAME + " WHERE " + NAME + " LIKE :name")
    Flowable<List<Art>> findByName(String name);

    @Query("SELECT * FROM " + TABLE_NAME + " WHERE (" + LATITUDE + " BETWEEN :minLat AND :maxLat) AND (" + LONGITUDE + " BETWEEN :minLon AND :maxLon)")
    Flowable<List<Art>> findInRegion(float maxLat, float minLat, float maxLon, float minLon);

    @Query("SELECT * FROM " + TABLE_NAME + " WHERE " + FAVORITE + " = 1 OR ((" + LATITUDE + " BETWEEN :minLat AND :maxLat) AND (" + LONGITUDE + " BETWEEN :minLon AND :maxLon))")
    Flowable<List<Art>> findInRegionOrFavorite(float maxLat, float minLat, float maxLon, float minLon);

    @Query("UPDATE " + TABLE_NAME + " SET " + FAVORITE + " = :isFavorite WHERE " + PLAYA_ID + " in (:playaIds)")
    void updateFavorites(List<String> playaIds, boolean isFavorite);

//    @Query("SELECT * FROM " + TABLE_NAME + " WHERE " + AUDIO_TOUR_URL + " IS NOT NULL")
//    Flowable<List<Art>> getAllWithAudioTour();

    @Insert
    void insert(Art... arts);

    @Update
    void update(Art... arts);
}

File: iBurn/src/main/java/com/gaiagps/iburn/database/ArtDao.kt
package com.gaiagps.iburn.database

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.Query
import androidx.room.Update
import kotlinx.coroutines.flow.Flow

/**
 * Created by dbro on 6/8/17.
 */
@Dao
interface ArtDao {
    @get:Query(
        "SELECT a.*, CASE WHEN f." + Favorite.PLAYA_ID +
            " IS NOT NULL THEN 1 ELSE 0 END AS " + UserData.FAVORITE +
            " FROM " + Art.TABLE_NAME + " a LEFT JOIN " + Favorite.TABLE_NAME +
            " f ON a." + PlayaItem.PLAYA_ID + " = f." + Favorite.PLAYA_ID +
            " ORDER BY " + PlayaItem.NAME
    )
    val all: Flow<List<ArtWithUserData>>
    @get:Query(
        "SELECT a.*, 1 AS " + UserData.FAVORITE +
            " FROM " + Art.TABLE_NAME + " a INNER JOIN " + Favorite.TABLE_NAME +
            " f ON a." + PlayaItem.PLAYA_ID + " = f." + Favorite.PLAYA_ID
    )
    val favorites: Flow<List<ArtWithUserData>>

    @Query(
        "SELECT a.*, CASE WHEN f." + Favorite.PLAYA_ID +
            " IS NOT NULL THEN 1 ELSE 0 END AS " + UserData.FAVORITE +
            " FROM " + Art.TABLE_NAME + " a LEFT JOIN " + Favorite.TABLE_NAME +
            " f ON a." + PlayaItem.PLAYA_ID + " = f." + Favorite.PLAYA_ID +
            " WHERE a." + PlayaItem.NAME + " LIKE :name"
    )
    fun findByName(name: String?): Flow<List<ArtWithUserData>>

    @Query(
        "SELECT a.*, CASE WHEN f." + Favorite.PLAYA_ID +
            " IS NOT NULL THEN 1 ELSE 0 END AS " + UserData.FAVORITE +
            " FROM " + Art.TABLE_NAME + " a LEFT JOIN " + Favorite.TABLE_NAME +
            " f ON a." + PlayaItem.PLAYA_ID + " = f." + Favorite.PLAYA_ID +
            " JOIN " + ArtFts.TABLE_NAME +
            " ON a." + PlayaItem.ID + " = " + ArtFts.TABLE_NAME + ".rowid" +
            " WHERE " + ArtFts.TABLE_NAME + " MATCH :query"
    )
    fun searchFts(query: String?): Flow<List<ArtWithUserData>>

    @Query(
        "SELECT a.*, CASE WHEN f." + Favorite.PLAYA_ID +
            " IS NOT NULL THEN 1 ELSE 0 END AS " + UserData.FAVORITE +
            " FROM " + Art.TABLE_NAME + " a LEFT JOIN " + Favorite.TABLE_NAME +
            " f ON a." + PlayaItem.PLAYA_ID + " = f." + Favorite.PLAYA_ID +
            " WHERE (a." + PlayaItem.LATITUDE + " BETWEEN :minLat AND :maxLat) " +
            "AND (a." + PlayaItem.LONGITUDE + " BETWEEN :minLon AND :maxLon)"
    )
    fun findInRegion(
        maxLat: Float,
        minLat: Float,
        maxLon: Float,
        minLon: Float
    ): Flow<List<ArtWithUserData>>

    @Query(
        "SELECT a.*, CASE WHEN f." + Favorite.PLAYA_ID +
            " IS NOT NULL THEN 1 ELSE 0 END AS " + UserData.FAVORITE +
            " FROM " + Art.TABLE_NAME + " a LEFT JOIN " + Favorite.TABLE_NAME +
            " f ON a." + PlayaItem.PLAYA_ID + " = f." + Favorite.PLAYA_ID +
            " WHERE f." + Favorite.PLAYA_ID + " IS NOT NULL OR ((a." + PlayaItem.LATITUDE +
            " BETWEEN :minLat AND :maxLat) AND (a." + PlayaItem.LONGITUDE +
            " BETWEEN :minLon AND :maxLon))"
    )
    fun findInRegionOrFavorite(
        maxLat: Float,
        minLat: Float,
        maxLon: Float,
        minLon: Float
    ): Flow<List<ArtWithUserData>>

    //    @Query("SELECT * FROM " + TABLE_NAME + " WHERE " + AUDIO_TOUR_URL + " IS NOT NULL")
    //    Flowable<List<Art>> getAllWithAudioTour();
    @Insert
    fun insert(vararg arts: Art?)

    @Update
    fun update(vararg arts: Art?)

    @Query(
        "SELECT a.*, CASE WHEN f." + Favorite.PLAYA_ID +
            " IS NOT NULL THEN 1 ELSE 0 END AS " + UserData.FAVORITE +
            " FROM " + Art.TABLE_NAME + " a LEFT JOIN " + Favorite.TABLE_NAME +
            " f ON a." + PlayaItem.PLAYA_ID + " = f." + Favorite.PLAYA_ID +
            " WHERE a." + PlayaItem.PLAYA_ID + " = :playaId"
    )
    fun findByPlayaId(playaId: String): Flow<ArtWithUserData>

    @Query(
        "SELECT a.*, CASE WHEN f." + Favorite.PLAYA_ID +
            " IS NOT NULL THEN 1 ELSE 0 END AS " + UserData.FAVORITE +
            " FROM " + Art.TABLE_NAME + " a LEFT JOIN " + Favorite.TABLE_NAME +
            " f ON a." + PlayaItem.PLAYA_ID + " = f." + Favorite.PLAYA_ID +
            " WHERE a." + PlayaItem.ID + " = :id"
    )
    fun findById(id: Int): Flow<ArtWithUserData>
}
